src/m365/tenant/commands/auditlog/auditlog-report.ts

```diff
@@ -0,0 +1,272 @@
+import commands from '../../commands';
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  contentType: string;
+  startTime?: string;
+  endTime?: string;
+}
+
+interface ActivityfeedSubscription {
+  contentType: string;
+  status: string;
+  webhook: string;
+}
+
+interface AuditContentList {
+  contentType: string;
+  contentId: string;
+  contentUri: string;
+  contentCreated: string;
+  contentExpiration: string;
+}
+
+interface AuditlogReport {
+  CreationTime: string;
+  Id: string;
+  Workload: string;
+  Operation: string;
+  UserId: string;
+}
+
+enum AuditContentTypes {
+  AzureActiveDirectory = "Audit.AzureActiveDirectory",
+  Exchange = "Audit.Exchange",
+  SharePoint = "Audit.SharePoint",
+  General = "Audit.General",
+  DLP = "DLP.All"
+}
+class TenantAuditlogReportCommand extends Command {
+  private serviceUrl: string = 'https://manage.office.com/api/v1.0';
+  private tenantId: string | undefined;
+  private batchedPromises: any = [];
+  private completeAuditReports: AuditlogReport[] = [];
+
+  public get name(): string {
+    return `${commands.TENANT_AUDITLOG_REPORT}`;
+  }
+
+  public get description(): string {
+    return 'Gets audit logs from the Office 365 Management API';
+  }
+
+  public getTelemetryProperties(args: CommandArgs): any {
+    const telemetryProps: any = super.getTelemetryProperties(args);
+    telemetryProps.startTime= typeof args.options.startTime !== 'undefined';
+    telemetryProps.endTime = typeof args.options.endTime !== 'undefined';
+    telemetryProps.contentType = args.options.contentType;
+    return telemetryProps;
+  }
+
+  public defaultProperties(): string[] | undefined {
+    return ['CreationTime', 'Operation', 'ClientIP', 'UserId', 'Workload'];
+  }
+
+  public commandAction(logger: Logger, args: CommandArgs, cb: (err?: any) => void): void {
+    if (this.verbose) {
+      logger.logToStderr(`Start retrieving Audit Log Report`);
+    }
+
+    this.tenantId = Utils.getTenantIdFromAccessToken(auth.service.accessTokens[auth.defaultResource].value);
+    this.getCompleteAuditReports(args, logger)
+      .then((res: AuditlogReport[]): void => {
+        logger.log(res);
+        if (this.verbose) {
+          logger.logToStderr(chalk.green('DONE'));
+        }
+
+        cb();
+      }, (err: any): void => this.handleRejectedODataJsonPromise(err, logger, cb));
+  }
+
+  private getCompleteAuditReports(args: CommandArgs, logger: Logger): Promise<AuditlogReport[]> {
+    return this.startContentSubscriptionifNotActive(args, logger)
+      .then((): Promise<AuditContentList[]> => this.getAuditContentList(args, logger))
+      .then((AuditContentLists: AuditContentList[]): void => this.getBatchedPromises(AuditContentLists, 10))
+      .then((): Promise<void> => { return new Promise<void>((resolve: () => void, reject: (error: any) => void): void => { if (this.batchedPromises.length > 0) this.getBatchedAuditlogData(logger, 0, resolve, reject); }) })
```

Comment:
Hey @waldekmastykarz  - Kindly find my comments below for your suggestion. Please feel free to let me know if my thoughts / comments does not make much sense in this context, I will be more than happy to alter that. 

> What if `this.batchedPromises` is `0`? In that case the Promise would end up unresolved?

That situation will come only when you don't have any ContentURL at all (which means there are no audit logs available for the selected ContentType). In that case, all we need to is to give empty resolve. Hence if we have even a single ContentURL, variable `this.batchedPromises` will be of length 1. Hence the case where the Promise will be unresolved will not occur. 

> If batchedPromises is always greater than 0, then we should remove the if statement and call getBatchedAuditlogData directly.

The case where `this.batchedPromises` will be 0 is when there are no audit logs available. So there will be hypothetical cases where there no ContentURL available since there are no Audit Logs available and hence we may have to give a condition check.

But I totally understood your thought process of not giving the condition check in the root, but give it in the function and handle accordingly. 
My though process on that was, imagine a situation where you have larger number of Content URLs (say 30). In that case, when we batch it, every time a condition check is needed since the function is a recursive function. Rather than a repeated process, I thought it would be better to handle the **No Content Case** in the root since we don't need a repeated check in the child recursive function. 
If you think we can handle it differently in an efficient way, I will super happy to implement that. 

Would be great if you can let me know your thoughts / comments on that

Reply:
What I'm trying to understand is how this promise will be completed if batchedPromises will be 0? Following the current code structure, we'd end up with an unresolved promise and a command that wouldn't stop running. If the case is handled elsewhere, then we should be able to safely assume that at this stage, batchedPromises will always be greater than 0, right?

Reply:
That is a big miss from my end. I had not realized that the Promises were not getting resolved when there are not audit logs and the command was still in running mode. Big big blunder from my end 😥. I can't thank the patience and dedication you possess in explaining this in detail even to the grassroot level. ❤

I have modified the condition ([Line 91](https://github.com/arjunumenon/cli-microsoft365/blob/18c5824a71523bb97acc33b4f173d9c2f3d2d5bd/src/m365/tenant/commands/auditlog/auditlog-report.ts#L98) to Line 92) and have given a resolve when there are no audit data and with that Promises will get resolved and will have a closure. Please feel free to let me know if that is not what you were looking for.